Civi/Funding/FundingCase/Handler/FundingCaseFinishClearingHandler.php

```diff
@@ -0,0 +1,116 @@
+
+  private FundingCaseActionsDeterminerInterface $actionsDeterminer;
+
+  private DrawdownManager $drawdownManager;
+
+  private FundingCaseManager $fundingCaseManager;
+
+  private PayoutProcessManager $payoutProcessManager;
+
+  private RequestContextInterface $requestContext;
+
+  private FundingCaseStatusDeterminerInterface $statusDeterminer;
+
+  public function __construct(
+    FundingCaseActionsDeterminerInterface $actionsDeterminer,
+    DrawdownManager $drawdownManager,
+    FundingCaseManager $fundingCaseManager,
+    PayoutProcessManager $payoutProcessManager,
+    RequestContextInterface $requestContext,
+    FundingCaseStatusDeterminerInterface $statusDeterminer
+  ) {
+    $this->actionsDeterminer = $actionsDeterminer;
+    $this->drawdownManager = $drawdownManager;
+    $this->fundingCaseManager = $fundingCaseManager;
+    $this->payoutProcessManager = $payoutProcessManager;
+    $this->requestContext = $requestContext;
+    $this->statusDeterminer = $statusDeterminer;
+  }
+
+  /**
+   * @throws \Civi\Funding\Exception\FundingException
+   * @throws \CRM_Core_Exception
+   */
+  public function handle(FundingCaseFinishClearingCommand $command): void {
+    $fundingCase = $command->getFundingCase();
+    $this->assertAuthorized($command);
+
+    $payoutProcess = $this->payoutProcessManager->getLastByFundingCaseId($fundingCase->getId());
+    Assert::notNull($payoutProcess);
+    $this->drawdownManager->deleteNewDrawdownsByPayoutProcessId($payoutProcess->getId());
+
+    $amountRemaining = $this->fundingCaseManager->getAmountRemaining($fundingCase->getId());
+    if (0.0 !== $amountRemaining) {
+      $drawdown = DrawdownEntity::fromArray([
+        'payout_process_id' => $payoutProcess->getId(),
+        'status' => 'new',
+        'creation_date' => date('Y-m-d H:i:s'),
+        'amount' => $amountRemaining,
+        'acception_date' => NULL,
+        'requester_contact_id' => $this->requestContext->getContactId(),
+        'reviewer_contact_id' => NULL,
+      ]);
+      $this->drawdownManager->insert($drawdown);
+      $this->drawdownManager->accept($drawdown, $this->requestContext->getContactId());
+    }
+
+    $this->payoutProcessManager->close($payoutProcess);
+
+    $fundingCase->setStatus($this->statusDeterminer->getStatus(
+      $fundingCase->getStatus(),
+      FundingCaseActions::FINISH_CLEARING
+    ));
+    $this->fundingCaseManager->update($fundingCase);
+  }
+
+  /**
+   * @throws \Civi\API\Exception\UnauthorizedException
+   */
+  private function assertAuthorized(FundingCaseFinishClearingCommand $command): void {
+    if (!$this->actionsDeterminer->isActionAllowed(
+      FundingCaseActions::FINISH_CLEARING,
+      $command->getFundingCase()->getStatus(),
+      $command->getApplicationProcessStatusList(),
+      $command->getFundingCase()->getPermissions(),
+    )) {
+      throw new UnauthorizedException(E::ts('Finishing the clearing of this funding case is not allowed.'));
```

Comment:
Perhaps add `$command->getFundingCase()->getIdentifier()` to exception message for better error analysis?

Reply:
Changed.